# Data Preprocessing Notebook

# Introduction

The initial dataset being sourced is from the HuggingFace library, (https://huggingface.co/datasets/brianarbuckle/cocktail_recipes)

This initial dataset contains columns `title`, `ingredients`, `directions`, `misc`, `source`, and `ner`.

**Overview**
The 'Ingredients' column of our dataset presents a unique challenge. It consists of objects, each being a list of strings, where each string details a part of a cocktail recipe. These details typically include the quantity, unit, and name of an ingredient, but they can also contain preparation instructions or garnishing details that are not directly usable in their current form for data analysis or recipe generation.

**Objective**
Our goal is to parse these strings to extract structured information that can be effectively utilized in our project. Specifically, we aim to separate the ingredient details into distinct components: quantity, unit, and ingredient name, while filtering out the non-ingredient related information.

## Imports

In [8]:
import pandas as pd
import re
from datasets import load_dataset

pip install datasetsto access HuggingFace datasets library

In [9]:
#!pip install datasets

# Initial Data Inspection

In [10]:
# Loading the cocktail_recipes dataset
dataset = load_dataset('brianarbuckle/cocktail_recipes')

In [11]:
df = pd.DataFrame(dataset['train'])

In [12]:
df.head(5)

title                                        ingredients  \
0     151 Swizzle  [1.5 oz. 151-Proof Demerara Rum [Lemon Hart or...   
1    20th Century  [The 21st Century, 2 oz. Siete Leguas Blanco T...   
2    20th Century  [1.5 oz. Plymouth Gin, 3\/4 oz. Mari Brizard W...   
3  Abbey Cocktail                                                 []   
4   Absinthe Drip  [1 1/2 ounces Pernod (or other absinthe substi...   

                                          directions  \
0                                                 []   
1  [shake on ice and strain into coupe , The Best...   
2                          [shake on ice and strain]   
3  [Shake liquid ingredients with ice., Strain in...   
4  [Pour Pernod into a pousse-cafŽ or sour glass....   

                                                misc                  source  \
0                                                 []  Beachbum Berry Remixed   
1                                                 []              Jim Meehan   
2                                                 []                     PDT   
3            [Suggested glassware is Cocktail Glass]   The Ultimate Bar Book   
4  [The Absinthe Drip was made famous by Toulouse...   The Ultimate Bar Book   

                                   ner  
0                        [pernod, rum]  
1  [cocchi americano, pernod, tequila]  
2                        [lillet, gin]  
3                                   []  
4                   [pernod, absinthe]

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 875 entries, 0 to 874
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        875 non-null    object
 1   ingredients  875 non-null    object
 2   directions   875 non-null    object
 3   misc         875 non-null    object
 4   source       875 non-null    object
 5   ner          875 non-null    object
dtypes: object(6)
memory usage: 41.1+ KB


In [29]:
df.shape

(875, 6)

From this **Initial Data Inspection**, we see that the data is a 875 x 6 dataframe, entirely of objects. We also get to see that there are multiple instances of entries being `[]` constiting non-null entries.

# Data Cleaning

Section involves removing empty rows, rows irrelevant to the analysis, or any rows unusable in the analysis.

In [14]:
# Removing rows where `ingredients` is empty
non_empty_ingredients_df = df[df['ingredients'].apply(lambda x: len(x) > 0)]
modified_df = non_empty_ingredients_df.reset_index(drop=True)

In [15]:
modified_df.shape

(873, 6)

From this, we can see that the dataframe went from 875 → 873 rows. So only two rows contained no information in the ingredients section. 

# Parsing 'Ingredients'

This section will involve parsing the 'Ingredients' column in order to obtain data relevant to the project, and organize it such that it can be utilized.

For example, the current 'ingredients' column contains data such as:

In [27]:
modified_df['ingredients'][0]

['1.5 oz. 151-Proof Demerara Rum [Lemon Hart or El Dorado]',
 '0.5 oz. Lime Juice',
 '0.5 oz. Sugar Syrup',
 '1 Dash Angostura Bitters',
 '6 drops [1\\/8 Tsp.] Pernod',
 '8 oz. crushed ice',
 'blend at high speed for 5 seconds, pour into metal cup or a pilsner glass',
 'Garnish with dust of nutmeg and a cinnamon stick']

## Analysis

- **Ingredient Entries**: Most strings follow a common format: `[quantity] [unit] [ingredient]`, which allows for standardized extraction. For example, `1.5 oz. 151-Proof Demerara Rum [Lemon Hart or El Dorado]` contains clear quantity, unit, and ingredient name components.

- **Non-Ingredient Instructions**: Some strings contain instructions or garnishing details, such as `blend at high speed for 5 seconds, pour into metal cup or a pilsner glass`, which do not fit the standard ingredient format and are thus considered unusable for direct ingredient analysis.


## Strategy
To tackle this, we employ a parsing function that utilizes regular expressions to identify and extract the quantity, unit, and ingredient name from each string, while discarding any non-ingredient information. This function is designed to handle the common format efficiently and includes error handling for strings that do not conform to the expected pattern.

By parsing the 'Ingredients' column in this manner, we aim to create a structured and usable dataset that will serve as a solid foundation for the subsequent stages of our project, including data analysis and the generation of new cocktail recipes.

In [33]:
def parse_ingredient(ingredient_str):
    # First, pre-process to correct known formatting issues with fractions
    corrected_str = re.sub(r'\\', '', ingredient_str)  # Remove backslashes that might interfere with parsing
    
    # regex patterns to extact ingredients from strings
    pattern = re.compile(
        r'(?P<quantity>\d+\s*\d*\/\d+|\d*\.\d+|\d+)?\s*'  # Capture quantities, fractions, decimals
        r'(?P<unit>oz|ounces?|tsp|teaspoons?|tablespoons?|tbl|tbs|cups?|pints?|quarts?|gallons?|lbs?|pounds?|ml|mL|liters?|dash|dashes|drops?|pinch|pinches)?\.?\s*'
        r'(?P<ingredient>.+)', re.IGNORECASE)

    match = pattern.match(corrected_str)
    
    if match:
        # Normalize unit names
        unit = match.group('unit')
        if unit:
            unit = unit.lower()
            if unit in ['tsp', 'teaspoons']:
                unit = 'tsp'
            elif unit in ['tbl', 'tbs', 'tablespoons']:
                unit = 'tbsp'
            elif unit in ['pounds', 'pound', 'lbs']:
                unit = 'lb'
            elif unit in ['gallons', 'gallon']:
                unit = 'gal'
            elif unit in ['ounces', 'ounce', 'oz']:
                unit = 'oz'
            elif unit in ['milliliters', 'millilitre', 'ml', 'mL']:
                unit = 'ml'
            elif unit in ['liters', 'litre']:
                unit = 'l'
        
        return {'quantity': match.group('quantity'), 'unit': unit, 'ingredient': match.group('ingredient').strip()}
    else:
        # Handling cases that don't fit the expected pattern
        return {'quantity': None, 'unit': None, 'ingredient': corrected_str}

In [18]:
parsed_ingredients_list = []

In [34]:
for index, row in modified_df.iterrows():
    for ingredient in row['ingredients']:
        parsed_ingredient = parse_ingredient(ingredient)
        parsed_ingredient['recipe_id'] = index  # Keep track of the recipe ID
        parsed_ingredients_list.append(parsed_ingredient)

parsed_df = pd.DataFrame(parsed_ingredients_list)

We have created `parsed_df` which contains the parsed versions of the `ingredients` column of the original dataset.

In [20]:
parsed_df.shape

(3722, 4)

In [21]:
parsed_df.head(50)

quantity   unit                                         ingredient  \
0       1.5     oz   151-Proof Demerara Rum [Lemon Hart or El Dorado]   
1       0.5     oz                                         Lime Juice   
2       0.5     oz                                        Sugar Syrup   
3         1   dash                                  Angostura Bitters   
4         6  drops                                  [1/8 Tsp.] Pernod   
5         8     oz                                        crushed ice   
6      None   None  blend at high speed for 5 seconds, pour into m...   
7      None   None   Garnish with dust of nutmeg and a cinnamon stick   
8      None   None                                   The 21st Century   
9         2     oz                        Siete Leguas Blanco Tequila   
10      3/4     oz                        Marie Brizard Crme de Cacao   
11      3/4     oz                                        Lemon Juice   
12     None   None                      rinse coupe glass with Pernod   
13      1.5     oz                                       Plymouth Gin   
14      3/4     oz                   Mari Brizard White Crme de Cacao   
15      3/4     oz                                       Lillet Blanc   
16      3/4     oz                                        Lemon Juice   
17    1 1/2     oz              Pernod (or other absinthe substitute)   
18        1   None                                         sugar cube   
19        5     oz                                cold filtered water   
20        1   None                                   perforated spoon   
21        1     oz                                       gold tequila   
22        1     oz                                           gold rum   
23        2     oz                             fresh grapefruit juice   
24        3     oz                                    pineapple juice   
25     1.25     oz                                             Brandy   
26      0.5     oz                                 Amontillado Sherry   
27      0.5     oz                                        Lemon Juice   
28        1   None                                Barspoon Lime Juice   
29      0.5     oz                               Demerara Syrup (2:1)   
30        1     oz                                              Water   
31     None   None                    mix all ingredients and add ice   
32     None   None                Lemon slice and fresh grated nutmeg   
33     None   None                                     2.5 cups Sugar   
34        1    cup                                              Water   
35       18     oz                                        Lemon Juice   
36        4     oz                                         Lime Juice   
37        1   None                                     ,500 mL Brandy   
38       18     oz                                    Montilla Sherry   
39      1.5   None                                          qt. Water   
40     None   None                            dissolve Sugar in Water   
41        4   None                                   Hogsheads Brandy   
42        8   None                                    Hogsheads Water   
43       25   None                                        ,000 Lemons   
44       20   None                                    Gal. Lime Juice   
45        1   None                           ,300 Weight Lisbon Sugar   
46        5     lb                                      Grated Nutmeg   
47      300   None                                    Toasted Biscuts   
48        1   None                                        Pipe Malaga   
49     None   None                            dissolve Sugar in Water   

    recipe_id  
0           0  
1           0  
2           0  
3           0  
4           0  
5           0  
6           0  
7           0  
8           1  
9           1  
10          1  
11          1  
12          1  
13          2  
14          2  
15          2  
16     

Looking into remaining edge cases, if something appears in the following, that means the quantity or unit is not successfully integrated by the parsing above.

In [23]:
# Filter rows where either 'quantity' or 'unit' is None (NaN)
none_rows = parsed_df.loc[parsed_df['quantity'].isnull() | parsed_df['unit'].isnull()]

# Display these rows
print(none_rows)

     quantity  unit                                         ingredient  \
6        None  None  blend at high speed for 5 seconds, pour into m...   
7        None  None   Garnish with dust of nutmeg and a cinnamon stick   
8        None  None                                   The 21st Century   
12       None  None                      rinse coupe glass with Pernod   
18          1  None                                         sugar cube   
...       ...   ...                                                ...   
3713     None  None                                  2oz. 90 Proof Rye   
3718     None  None                           shake on ice and strain"   
3719     None  None                                          champagne   
3720     None  None                                   orange juice-ish   
3721     None  None                  optional but preferable 1 oz. gin   

      recipe_id  
6             0  
7             0  
8             1  
12            1  
18            3  
...